pints/_mcmc/_slice_hyperrectangles.py

```diff
@@ -0,0 +1,283 @@
+# -*- coding: utf-8 -*-
+#
+# Hyperrectangles-based Slice Sampling
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017-2019, University of Oxford.
+#  For licensing information, see the LICENSE file distributed with the PINTS
+#  software package.
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import pints
+import numpy as np
+
+
+class SliceHyperrectanglesMCMC(pints.SingleChainMCMC):
+    """
+    Implements Hyperrectangles-based Slice Sampling, as described in [1].
+
+    This is a multivariate method, which generates n-dimensional samples of
+    the form ``x = (x_1, ..., x_n)`` by sampling uniformly from the area of an
+    axis-aligned hyperrectangle:
+    ``H = {x: L_i < x_i < R_i for all i = 1, ..., n}``.
+    Here, ``L_i`` and ``R_i`` define  the extent of the hyperrectangle along
+    the ``i`` th axis.
+
+    Sampling follows:
+
+    1. Calculate the pdf (``f(x0)``) of the current sample (``x0``).
+    2. Draw a real value (``y``) uniformly from (0, f(x0)), defining a
+       horizontal “slice”: S = {x: y < f (x)}. Note that ``x0`` is
+       always within S.
+    3. Find a hyperrectangle (``H = (L_1, R_1) ×···× (L_n, R_n)``) around
+       ``x_0``, which preferably contains at least a big part of the slice.
+    4. Draw a new point (``x1``) from the part of the slice within this
```

Comment:
Is rejection sampling used for step 4? For example, if a point is drawn that is in the rectangle not in the slice, then it is rejected, right? Can we say this explicitly?